File: go/gapcore/signature.go
package gapcore

import (
	"crypto/ed25519"
	"errors"
	"strconv"
)

// PAE builds the DSSE (Dead Simple Signing Envelope) Pre-Authentication
// Encoding used by GAP attestations:
//
//	PAE(type, body) = "DSSEv1" SP LEN(type) SP type SP LEN(body) SP body
//
// where LEN is the ASCII decimal BYTE length. Binding payloadType and both
// lengths into the signed bytes is what makes a signature minted for one
// payloadType fail against another. Matches @synoi/sraid attestation.pae.
func PAE(payloadType, payload string) []byte {
	typeBytes := []byte(payloadType)
	bodyBytes := []byte(payload)
	prefix := "DSSEv1 " + strconv.Itoa(len(typeBytes)) + " " + payloadType + " " + strconv.Itoa(len(bodyBytes)) + " "
	out := make([]byte, 0, len(prefix)+len(bodyBytes))
	out = append(out, prefix...)
	out = append(out, bodyBytes...)
	return out
}

// VerifyAttestationEd25519 verifies the ed25519 signature leg of a DSSE
// attestation over PAE(payloadType, payload). The GAP attestation profile is
// hybrid (ed25519 AND ml-dsa-65 both required); this verifies only the ed25519
// leg. See README "Post-quantum status" for the ml-dsa-65 leg.
func VerifyAttestationEd25519(pub []byte, payloadType, payload string, sig []byte) bool {
	return VerifyDetached(pub, PAE(payloadType, payload), sig)
}

// SignDetached produces a detached Ed25519 signature over message using the
// given seed (32-byte private key / RFC 8032 seed). The signature is the
// standard 64-byte Ed25519 detached signature, matching the TS reference's
// ed25519.sign (@noble/curves) over the same message bytes.
func SignDetached(seed []byte, message []byte) ([]byte, error) {
	if len(seed) != ed25519.SeedSize {
		return nil, errors.New("gapcore: ed25519 seed must be 32 bytes")
	}
	priv := ed25519.NewKeyFromSeed(seed)
	return ed25519.Sign(priv, message), nil
}

// PublicFromSeed derives the 32-byte Ed25519 public key from a 32-byte seed.
func PublicFromSeed(seed []byte) ([]byte, error) {
	if len(seed) != ed25519.SeedSize {
		return nil, errors.New("gapcore: ed25519 seed must be 32 bytes")
	}
	priv := ed25519.NewKeyFromSeed(seed)
	pub := priv.Public().(ed25519.PublicKey)
	return pub, nil
}

// VerifyDetached verifies a detached Ed25519 signature over message against a
// 32-byte public key. Returns false (never panics) on malformed inputs, so it
// is safe to call directly on untrusted conformance-vector data.
func VerifyDetached(pub []byte, message []byte, sig []byte) bool {
	if len(pub) != ed25519.PublicKeySize || len(sig) != ed25519.SignatureSize {
		return false
	}
	return ed25519.Verify(ed25519.PublicKey(pub), message, sig)
}

// SignCanonical canonicalizes value (COF profile), then produces a detached
// Ed25519 signature over the canonical UTF-8 bytes. This is the single-key
// self-custody signing surface mirrored from receipt.ts.
func SignCanonical(seed []byte, value interface{}) (canonical string, sig []byte, err error) {
	canonical, err = Canonicalize(value)
	if err != nil {
		return "", nil, err
	}
	sig, err = SignDetached(seed, []byte(canonical))
	if err != nil {
		return "", nil, err
	}
	return canonical, sig, nil
}

File: rust/src/signature.rs
//! Signature primitives.
//!
//! Two surfaces, matching the TypeScript reference:
//!
//!   * Ed25519 detached sign + verify over canonical bytes. This is the
//!     self-custody single-key receipt signing surface of `@synoi/gap`
//!     `receipt()` / `verifyReceiptSignature` (src/receipt.ts). We can both
//!     PRODUCE and VERIFY here; the conformance suite proves a sign->verify
//!     round trip and independently verifies the golden Ed25519 leg of the
//!     `sraid/signatures.json` vectors.
//!
//!   * Hybrid Ed25519 + ML-DSA-65 verify, `valid` only when BOTH legs verify,
//!     over the same canonical bytes. This is the `@synoi/sraid`
//!     `verifySignature` surface (src/signature.ts). The message is the
//!     canonical string encoded as UTF-8, signed directly (NOT pre-hashed),
//!     with an empty signing context. Verified byte-for-byte against the
//!     `@noble/post-quantum` ml_dsa65 golden signatures in
//!     `sraid/signatures.json`.

use base64::{engine::general_purpose::STANDARD, Engine};
use ed25519_dalek::{
    Signature as EdSignature, Signer, SigningKey, Verifier as EdVerifier, VerifyingKey,
};
use ml_dsa::signature::Verifier as MlVerifier;
use ml_dsa::{EncodedVerifyingKey, MlDsa65, Signature as MlSignature, VerifyingKey as MlVerifyingKey};

/// Sign `message` bytes with a 32-byte Ed25519 seed (private key), returning
/// the 64-byte detached signature. Mirrors `ed25519.sign(msg, priv)` in the
/// `@synoi/gap` receipt signer.
pub fn ed25519_sign(seed: &[u8; 32], message: &[u8]) -> [u8; 64] {
    let signing = SigningKey::from_bytes(seed);
    let sig: EdSignature = signing.sign(message);
    sig.to_bytes()
}

/// Derive the 32-byte Ed25519 public key from a 32-byte seed.
pub fn ed25519_public_key(seed: &[u8; 32]) -> [u8; 32] {
    SigningKey::from_bytes(seed).verifying_key().to_bytes()
}

/// Verify a 64-byte Ed25519 signature over `message` against a 32-byte public
/// key. Returns false on any malformed input rather than panicking, matching
/// the reference verifier's clean boolean contract.
pub fn ed25519_verify(sig: &[u8], message: &[u8], public_key: &[u8]) -> bool {
    let pk: [u8; 32] = match public_key.try_into() {
        Ok(a) => a,
        Err(_) => return false,
    };
    let sig_arr: [u8; 64] = match sig.try_into() {
        Ok(a) => a,
        Err(_) => return false,
    };
    let vk = match VerifyingKey::from_bytes(&pk) {
        Ok(v) => v,
        Err(_) => return false,
    };
    let signature = EdSignature::from_bytes(&sig_arr);
    vk.verify(message, &signature).is_ok()
}

/// Verify a raw ML-DSA-65 signature (3309 bytes) over `message` against a raw
/// ML-DSA-65 public key (1952 bytes), empty context. Returns false on any
/// malformed input. Byte-compatible with `@noble/post-quantum` ml_dsa65 and
/// with the OpenSSL 3.5 native path used by `@synoi/sraid`.
pub fn ml_dsa65_verify(sig: &[u8], message: &[u8], public_key: &[u8]) -> bool {
    let ek: EncodedVerifyingKey<MlDsa65> = match public_key.try_into() {
        Ok(e) => e,
        Err(_) => return false,
    };
    let vk = MlVerifyingKey::<MlDsa65>::decode(&ek);
    let sig_arr: [u8; 3309] = match sig.try_into() {
        Ok(a) => a,
        Err(_) => return false,
    };
    let signature = match MlSignature::<MlDsa65>::decode(&sig_arr.into()) {
        Some(s) => s,
        None => return false,
    };
    vk.verify(message, &signature).is_ok()
}

/// A hybrid signature envelope: base64-encoded Ed25519 and ML-DSA-65
/// signatures plus a signer key id. Mirrors `@synoi/sraid` `SignatureEnvelope`.
pub struct SignatureEnvelope {
    pub ed25519: String,
    pub ml_dsa_65: String,
    pub signer_kid: String,
}

/// Result of a hybrid verify. `valid` is true only when BOTH legs verify.
pub struct VerifyResult {
    pub valid: bool,
    pub reasons: Vec<String>,
}

/// Hybrid verify: `valid` only when the Ed25519 AND the ML-DSA-65 signatures
/// both verify against the supplied public keys and canonical bytes. Reasons
/// accumulate across both legs in one pass. Counterpart of `@synoi/sraid`
/// `verifySignature` (src/signature.ts).
pub fn verify_hybrid(
    canonical: &[u8],
    envelope: &SignatureEnvelope,
    ed25519_pub: &[u8],
    ml_dsa_pub: &[u8],
) -> VerifyResult {
    let mut reasons: Vec<String> = Vec::new();

    let ed_sig = match STANDARD.decode(&envelope.ed25519) {
        Ok(b) => Some(b),
        Err(_) => {
            reasons.push("ed25519-malformed".to_string());
            None
        }
    };
    let ml_sig = match STANDARD.decode(&envelope.ml_dsa_65) {
        Ok(b) => Some(b),
        Err(_) => {
            reasons.push("ml-dsa-malformed".to_string());
            None
        }
    };

    let mut ed_ok = false;
    let mut ml_ok = false;

    if let Some(ref s) = ed_sig {
        ed_ok = ed25519_verify(s, canonical, ed25519_pub);
        if !ed_ok {
            reasons.push("ed25519-invalid".to_string());
        }
    }
    if let Some(ref s) = ml_sig {
        ml_ok = ml_dsa65_verify(s, canonical, ml_dsa_pub);
        if !ml_ok {
            reasons.push("ml-dsa-invalid".to_string());
        }
    }

    VerifyResult {
        valid: ed_ok && ml_ok,
        reasons,
    }
}
